Google Keep Import Pull Request

daledesilva opened the pull request:
Please note that I have included a partial Google Keep export in the repository for easy testing. This means json files, and binary files can be ignored during code review.

I've also push the edges of the existing modal layout so that I could provide the user relevant information and options when importing. Screenshot included below.

<img width="606" alt="Screenshot 2023-08-01 at 4 49 53 pm" src="https://github.com/obsidianmd/obsidian-importer/assets/18710973/e2caef6c-71b4-4c83-8b8b-c09d3f326979">

daledesilva commented:
Testing data is also zipped here, but please also test any of your own exports from Google Keep.
[testing-data.zip](https://github.com/obsidianmd/obsidian-importer/files/12225487/testing-data.zip)

daledesilva commented:
FYI. I just realised I didn't adhere to the requirement of "**Follow the user's setting for Default location for new attachments**"
I'll modify that to update the PR soon and post here.

daledesilva commented:
I've made a suggestion on the [original ticket](https://github.com/obsidianmd/obsidian-importer/issues/12#issuecomment-1660145980) where I advise we don't **Follow the user's setting for default attachment location**
So I'll consider this this PR ready for review unless someone makes a call otherwise.

polyipseity commented:
> FYI. I just realised I didn't adhere to the requirement of "**Follow the user's setting for Default location for new attachments**" I'll modify that to update the PR soon and post here.

In case you do have to follow the setting, there is an undocumented method `Vault.getAvailablePathForAttachments` that returns an available attachment path following the setting: https://github.com/obsidianmd/obsidian-importer/blob/9d5a2a9e1be3bb2ee7f66b2d05a9c9ef52af5b59/formats/html.ts#L245-L246

(I am guessing that slight use of undocumented methods are okay here.)

lishid commented:
Could we move the testing data to a separate PR, and put that in a `/tests/keep/` folder instead of mixing with the other sources?

daledesilva commented:
I've now updated the PR to support the recent changes to how file picking works.
I've also moved the testing files to [this PR](https://github.com/obsidianmd/obsidian-importer/pull/40) which should be reviewed after this one is merged.

**This is now ready to be reviewed.**
Since the testing files aren't included in here, you can refer to the first post above to download those testing files.

lishid commented on util.ts:
```diff
@@ -1,18 +1,139 @@
 	return array.join('');
 }
+
+export function pathToFilename(path: string) {
```
This is no longer needed.

lishid commented on util.ts:
```diff
@@ -1,18 +1,139 @@
+		normalizedPath = '/';
+	}
+
+	const folder = this.app.vault.getAbstractFileByPath(normalizedPath);
```
`this` has no reference.

lishid replied:
Hmm seems like a lot of these are like that - could you pass `vault` in as a parameter?

daledesilva replied:
Done. Good pickup. Weird that it even worked.

lishid commented on util.ts:
```diff
@@ -1,18 +1,139 @@
+const ILLEGAL_FILENAME_CHARACTERS = '\\/:*?<>\"|';
+const ILLEGAL_FILENAME_RE = new RegExp('[' + escapeRegex(ILLEGAL_FILENAME_CHARACTERS) + ']', 'g');
+
+const ILLEGAL_HASHTAG_CHARACTERS = '\\:*?<>\"|!@#$%^&()+=\`\'~;,.';
```
I would prefer if we consistently name these `tag` instead of mixing in `hashtag`

lishid commented on util.ts:
```diff
@@ -1,18 +1,139 @@
+export async function addAliasToFrontmatter(alias: string, fileRef: TFile) {
+	const sanitizedAlias = alias.split('\n').join(', ');
+	await this.app.fileManager.processFrontMatter(fileRef, (frontmatter: any) => {      
+		if(!frontmatter['alias']) {
```
We now recommend the key `aliases` instead of `alias`.

daledesilva replied:
Should tag still be singular or plural too?

lishid replied:
Should be `tags`.

daledesilva commented:
All the above comments are now addressed 🙂

ericaxu commented:
Hi there, can you please take a look at the CLA agreement located [here](https://github.com/obsidianmd/obsidian-releases/blob/master/cla.md)? This is one of the steps needed before we can merge your change.

If you agree to it, please reply with "I have read the CLA agreement and I hereby sign the CLA".

daledesilva commented:
Thanks. I can see that, but I don't really no how to reply. There's no comment fields.

kepano commented:
> Thanks. I can see that, but I don't really no how to reply. There's no comment fields.

you can reply right here in the PR comments 😊

daledesilva commented:
oh.. ha ha. Whoops.

I have read the CLA agreement and I hereby sign the CLA.

lishid commented on util.ts:
```diff
@@ -1,18 +1,126 @@
+ * Searches a string for characters unsupported by Obsidian in the hashtag body and returns a snaitized string.
+ * If the # symbol is included at the start or anywhere else it will be removed.
+ */
+export function sanitizeHashtag(name: string): string {
```
Also remember to rename `Hashtag` here and below.

lishid commented on util.ts:
```diff
@@ -1,18 +1,126 @@
+ * Retrieves a reference to a specific folder in a vault. Creates it first if it doesn't exist.
+ * Must pass in the relevant vault.
+ */
+export async function getOrCreateFolder(folderPath: string, vault: Vault): Promise<TFolder> {
```
I've made a helper function to cover this use case in FormatImporter that should help! (latest master)

lishid commented on formats/keep/add-keep-frontmatter.ts:
```diff
@@ -0,0 +1,37 @@
+
+
+
+export async function addKeepFrontMatter(fileRef: TFile, keepJson: KeepJson, fileManager: FileManager) {
```
This is quite inefficient because it keeps re-reading the file from disk, writing it back. Can we wrap it all in a single `processFrontMatter` call, and just pass the frontmatter object to each helper?

daledesilva replied:
Will do :)

daledesilva commented on formats/keep/models/KeepJson.ts:
```diff
@@ -0,0 +1,50 @@
+export interface KeepListItem {
```
Sure.
I was keeping consistent with the models in the yarle folder. Happy to change it.

lishid replied:
Good point, I will come back and fix those later!

daledesilva commented:
All of the above is addressed.
This is ready for further review.

lishid commented on formats/keep-json.ts:
```diff
@@ -0,0 +1,202 @@
+import { addAliasToFrontmatter, addTagToFrontmatter, toSentenceCase } from "../util";
+import { PickedFile, PickedFolder, parseFilePath } from "filesystem";
+import { BlobWriter, Entry, TextWriter } from "@zip.js/zip.js";
+import { Transform } from "stream";
```
This is unused.

daledesilva replied:
Fixed and pushed.

lishid commented on formats/keep-json.ts:
```diff
@@ -0,0 +1,202 @@
+		
+				} catch(e) {
+					console.error(`${file.name} ::: ${curInnerFilename ?? curInnerFilename}`, e);
+					progress.reportFailed(`${file.name} / ${curInnerFilename}`, e);
```
Prefer to remove the extra spaces here around the `/`.

daledesilva replied:
I'm curious why. This is about readability. The spaces make it much easier to identify the containing zip file and the inner file separately.

daledesilva replied:
Done & pushed.

lishid commented on formats/keep-json.ts:
```diff
@@ -0,0 +1,202 @@
+					// else: Silently skip any other unsupported files in the zip
+		
+				} catch(e) {
+					console.error(`${file.name} ::: ${curInnerFilename ?? curInnerFilename}`, e);
```
Let's avoid adding extra console messages since they are already reported in `reportFailed`.

daledesilva replied:
Fixed and Pushed.

lishid commented on formats/keep-json.ts:
```diff
@@ -0,0 +1,202 @@
+	async copyFile(arrayBuffer: ArrayBuffer, folderPath: string, filename: string, progress: ProgressReporter) {
+		let assetFolder = await this.createFolders(folderPath);
+		await this.vault.createBinary(`${assetFolder.path}/${filename}`, arrayBuffer);
+		progress.reportNoteSuccess(filename);
```
This should be `reportAttachmentSuccess`

lishid commented on main.ts:
```diff
@@ -149,13 +154,15 @@ export class ImporterModal extends Modal {
 					el.addEventListener('click', async () => {
 						let progress = new ProgressReporter();
 						this.modalEl.addClass('is-loading');
+						new Notice('Import started.');
```
Let's not add stuff to the main script

daledesilva replied:
Fixed and pushed.
FYI. This was added because if the user is uploading a large number of files, it will appear to them that the program has hung - since there is no message and interface stays the same.

There is now https://github.com/obsidianmd/obsidian-importer/issues/35, but until that's implemented the issue will persist. Hence I implemented it in main as a placeholder because it shouldn't be implemented individually by importers.

lishid commented on util.ts:
```diff
@@ -21,6 +54,46 @@ export function genUid(length: number): string {
+
+/**
+ * Adds a single tag to the tag property in frontmatter and santises it.
+ * Must pass in app.fileManager.
```
This comment isn't accurate anymore.

daledesilva replied:
Fixed and pushed

lishid commented on util.ts:
```diff
@@ -21,6 +54,46 @@ export function genUid(length: number): string {
+		frontmatter['aliases'] = [sanitizedAlias];
+	} else {
+		if (!Array.isArray(frontmatter['aliases'])) {
+			frontmatter['aliases'] = frontmatter['aliases'].split(' ');
```
This should split by comma. Actually we should be using obsidian provided functions here for `parseFrontMatterTags` and `parseFrontMatterAliases`.

daledesilva replied:
Fixed and pushed

lishid commented on formats/keep/models/keep-json.ts:
```diff
@@ -0,0 +1,50 @@
+    //
+	title?: string;
+	textContent?: string;
+	listContent?: Array<KeepListItem>;
```
Prefer to use `Type[]` instead of `Array<Type>`.

daledesilva replied:
Fixed and pushed

lishid commented on formats/keep/convert-json-to-md.ts:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    if(jsonContent.attachments) {
+        for(let i=0; i<jsonContent.attachments.length; i++) {
```
Use a for-of loop here instead?

daledesilva replied:
Fixed and pushed

lishid commented:
Does google keep export in zip format directly? What's the benefit of supporting flat folder of unzipped files? I'm thinking we should probably keep it zip-only because the file selector accepts a lot of files right now and it's quite some clutter.

daledesilva commented:
> Does google keep export in zip format directly? What's the benefit of supporting flat folder of unzipped files? I'm thinking we should probably keep it zip-only because the file selector accepts a lot of files right now and it's quite some clutter.

Google Keep doesn't let you selectively export notes, so you'll always get a zip of all your keep notes even if you only want some of them. For that reason, I think allowing users to selectively import is useful.

But I also don't think we should limit the functionality that's already built and working here simply because of a stylistic limitation of the dialog (Which will continue to create issues with other importers too). I mentioned this [here](https://github.com/obsidianmd/obsidian-importer/issues/35#issuecomment-1659423225).

You used the term "flat folder" though, which leads to believe that maybe you mean selecting the files by folder? I actually never intentionally allowed that. I used the method **addFileChooserSetting** back when it only gave me an individual file chooser, but at some point in pulling the master branch that method has changed to include both file and folder buttons.

daledesilva commented:
I've completed all but one edit above. Happy to do it but just thought I'd draw attention to the discussion first.
https://github.com/obsidianmd/obsidian-importer/pull/37#discussion_r1287733847

lishid commented:
Ah yes, so there's now an optional flag you can pass to indicate whether you allow multiple files to be chosen.

lishid commented:
> Google Keep doesn't let you selectively export notes, so you'll always get a zip of all your keep notes even if you only want some of them. For that reason, I think allowing users to selectively import is useful.

One thing I'm not understanding is allowing selection of everything including attachments, if you can only select single files to import, it doesn't make much sense to select single attachment files no?

daledesilva commented:
> One thing I'm not understanding is allowing selection of everything including attachments, if you can only select single files to import, it doesn't make much sense to select single attachment files no?

I'm not sure I'm interpreting what you're saying correctly.

You can select multiple files to import, therefore, the user would ideally select a json file or multiple and their related attachment at the same time, but even if they don't and they notice missing attachments later, they can rerun the importer and just select those attachments too.

Or, alternatively, they can just select the zip, or even multiple zips.

lishid commented:
Hmm I was replying to this part:

> You used the term "flat folder" though, which leads to believe that maybe you mean selecting the files by folder? I actually never intentionally allowed that. I used the method addFileChooserSetting back when it only gave me an individual file chooser, but at some point in pulling the master branch that method has changed to include both file and folder buttons.

Selecting the container folder would end up being the same as selecting all the files in that folder.

daledesilva commented:
> Selecting the container folder would end up being the same as selecting all the files in that folder.

Yes it is (apart from it being recursive). That's true with any importer.

I'm not sure if there's a change here you'd like. If it's to remove the folder option, that makes sense, but I'm already using addFileChooserSetting, does that now support options?

lishid commented on util.ts:
```diff
@@ -1,8 +1,12 @@
 let reservedRe = /^\.+$/;
 let windowsReservedRe = /^(con|prn|aux|nul|com[0-9]|lpt[0-9])(\..*)?$/i;
 let windowsTrailingRe = /[\. ]+$/;
+let potentialTagsRe = /(#[^ ^#]*)/g; // Finds any non-whitespace sections starting with #
```
Let's not mix these into the file name sanitizer code.

lishid replied:
Actually, can we move all of these util functions into `formats/keep/util.ts`?

daledesilva replied:
Done & pushed.

lishid commented on formats/keep/models/keep-json.ts:
```diff
@@ -0,0 +1,50 @@
+export interface KeepListItem {
```
Can we move this up to `/formats/keep/models.ts`?

lishid commented on formats/keep/convert-json-to-md.ts:
```diff
@@ -0,0 +1,33 @@
+ * Reads a Google Keep JSON file and returns a markdown string.
+ */
+export function convertJsonToMd(jsonContent: KeepJson): string {
+    let mdContent = '';
```
Let's make this into an array and use `join('\n')` at the end to avoid multiple string concatenations.

daledesilva replied:
Fixed & pushed

lishid commented on formats/keep-json.ts:
```diff
@@ -0,0 +1,199 @@
+	async importKeepNote(rawContent: string, folder: TFolder, title: string, progress: ProgressReporter) {
+		let keepJson = convertStringToKeepJson(rawContent);
+		if(!keepJson) {
+			progress.reportFailed(`${title}.json (Invalid format)`, 'NOT A GOOGLE KEEP JSON');
```
Can we update these to properly use a filename for the first parameter?

daledesilva replied:
Was going to flag this with you actually @lishid 
I'll change this, but please note that the ProgressReporter currently accepts a reason, but doesn't display it anywhere. Without that the user will have no idea why something was skipped, which will caused confusion.

Also, Are you happy with the format of the reasons specified in the second parameters?
I can't remember whether I put them in all caps for a reason.

lishid replied:
That's right. ProgressReporter is a work-in-progress that I have setup for the backend API - we are working in parallel on the interface to display the progress as well as the reason/details.

Everything will be user-visible in the end; I would prefer if they were just regular prose instead of allcaps.
Also the file path might end up getting parsed for better formatting of containing folder vs filename itself, so that's why it's important to use a proper filename instead of adding debug information into it.

daledesilva replied:
Makes sense, that was my expectation but just wanted to flag it.

lishid reviewed (commented):
Ok just these and fixing the merge conflict and we should be good to go.

daledesilva commented:
All updated

lishid commented on util.ts:
```diff
@@ -13,12 +14,29 @@ export function sanitizeFileName(name: string) {
 		.replace(windowsTrailingRe, '');
 }
 
-export function genUid(length: number): string {
```
Hmm this shouldn't have been moved...

lishid reviewed (commented):
Could you also run the auto formatter `npm run lint` to fix the inconsistent spaces?
